### PR TITLE
[WIP] Add async settings loading and Jetpack datastore support

### PR DESCRIPTION
**Pull request by pelmenstar1**
- Async setting loading means that the operation can take some time and result in an error. So I added load-states (`Loading`, `Error`, and `Success`).
- Also settings are loaded in snapshots because it would be strange if a single setting had its own load state. Besides, it's easier to track the state when snapshots are used.

To add:
- [ ] Screens in demo
- [ ] Controls support
- [x] Loading indicator
- [x] Error handling

### CODE REVIEW

File: compose-settings-storage-datastore/src/main/java/com/alorma/compose/settings/storage/datastore/DataStoreAsyncImpl.kt [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    fun string(defaultValue: String) = result(defaultValue)
+    fun stringSet(defaultValue: Set<String>) = result(defaultValue)
+
+    @Suppress("UNCHECKED_CAST")
+    private fun<T> result(defaultValue: T): Result<T> {
+        return Result(intPreferencesKey(name) as Preferences.Key<T>, defaultValue)
+    }
+}
+
+inline fun<T> dataStoreSnapshotKey(name: String, resolver: DataStoreSnapshotKeySelector.() -> DataStoreSnapshotKeySelector.Result<T>): DataStoreSnapshot.Key<T> {
```
Comment (pelmenstar1) [inline compose-settings-storage-datastore/src/main/java/com/alorma/compose/settings/storage/datastore/DataStoreAsyncImpl.kt:118]: I haven't decided which version of key creation is better to use (one with `dataStore*type*settingsSnapshotKey` or `dataStoreSnapshotKey` builder). @alorma What do you think?

**Comment by alorma** [comment]
Hi @pelmenstar1 

Beside I love the Idea of the async preference, I'm not sure on merging it also with the data store implementation.

So, I've decided to check it and got an approach to use datastore in a synchronous way, I've made the work here:

https://github.com/alorma/Compose-Settings/pull/45

Could you check it and give me your thought?

That's just a first implementation, later on it could came up with your async solution. But I think it's a good approach for a first datastore impl.

**Comment by pelmenstar1** [comment]
@alorma  Ok, no problem.

BTW, I don't think the use of Jetpack DataStore in a synchronous context is a good idea, because:
- the disk IO can be really slow
- no error handling

**Comment by pelmenstar1** [comment]
I'm closing the PR as the implementation of Jetpack DataStore support, which better fits in the current state of the library, has been merged.

**Pull request closed without merging**